### DISCUSSION TITLE
Carousel excess memory consumption

### DISCUSSION THREAD

**Discussion by vo1s** [AvaloniaUI/Avalonia#13931] [Q&A] [answered]
I was using this [sample](https://github.com/AvaloniaUI/Avalonia/blob/master/samples/ControlCatalog/Pages/CarouselPage.xaml) to create a carousel in avalonia. The only thing I changed was that I binded to ObservableCollection<Image>. Later I discovered that I must not use the Image in viewModel because it breaks mvvm. So I had this code:
Carousel Styled property to bind to observableCollection in carouselPage code-behind:
```
public static readonly StyledProperty<IEnumerable > ItemSourcePrProperty = AvaloniaProperty.Register<CarouselPage, IEnumerable >(
        nameof(ItemSourcePr));

    public IEnumerable  ItemSourcePr
    {
        get => GetValue(ItemSourcePrProperty);
        set => SetValue(ItemSourcePrProperty, value);
    }
```
And bind it to carousel ItemSource:
```
 <Carousel Name="carousel"  ItemsSource="{Binding ItemSourcePr, ElementName=CoureselControl}" Grid.Row="1">
```
Then I use this control in mainWindow like this:
```
<control:CarouselPage ItemSourcePr="{Binding MainImageCollection}" ... ></control:CarouselPage>
```
where MainImageCollection was `[ObservableProperty] private ObservableCollection<Image> _mainImageCollection = new() { };`
So on some action I was changing the _mainImageCollection like this:
```
MainImageCollection.Clear();
var files = FileManagerService.GetPlateNumber(PathToCurrentCarFolder);
            foreach (var s in files)
            {
                MainImageCollection.Add(new Image{ Source = new Bitmap(s)});
            }
```
And everything worked fine until a mention that every time I changed the collection I have my RAM memory consumption growing. Seems like old Images are staying in memory. And for example there 130 photos in there. And memory consumption is becoming 1.5gb. Then I changed the collection to 5 photos. But memory still 1.5gb or a bit more.

I know that I have to use only Bitmap without Image control. But Carousel doesn't show anything if I use it without Image.

**Comment by timunie** [comment]
Bitmap intead of Image. Use DataTemplate to show the image actually.

**Reply by timunie** [reply]
ItemTemplate should be there for Carousel, no?

**Reply by vo1s** [reply]
Like this?
```
<Carousel Name="imageCarousel">
    <Carousel.DataTemplates>
        <DataTemplate>
            <Image Source="{Binding}" Stretch="Uniform" Width="200" Height="150"/>
        </DataTemplate>
    </Carousel.DataTemplates>
</Carousel>

```

**Reply by timunie** [reply]
```xml
<Carousel ItemsSource="{Binding Items}" >
	<Carousel.ItemTemplate>
		<DataTemplate DataType="system:String">
			<Button Content="{Binding}"></Button>
		</DataTemplate>
	</Carousel.ItemTemplate>
</Carousel>
```

**Reply by vo1s** [reply]
Why is there a button? sorry, didn't get it. I just need somehow to properly bind to ObservableColection of photos. Because I need to scroll them in carousel

**Reply by timunie** [reply]
this is just some dummy data. You have to adjust the DataTemplate to your needs, so an image bound to the Bitmap

**Reply by vo1s** [reply]
I don't think that's a problem. Before that, i just used `<Carousel Name="carousel"  ItemsSource="{Binding ItemSourcePr, ElementName=CoureselControl}"/>` ItemSource property and that worked fine without templates. But if I use DataTemplate with just 1 image inside, will it be showing just 1 images, instead of all of them?

**Reply by thevortexcloud** [reply]
Item templates are applied per item. Meaning it will be applied to every item in the collection. It's just a way to explain to Avalonia how the item should be presented. 

The list box documentation explains this as well.

https://docs.avaloniaui.net/docs/reference/controls/listbox#item-template

Pretty much all collection controls work fairly similar to each other in terms of how you use them.

**Reply by thevortexcloud** [reply]
You bind to the `ItemsSource`. EG

```xaml
<Carousel Name="carousel" Grid.Row="1" ItemsSource="{Binding MyItemCollection}">
```

**Reply by vo1s** [reply]
Ok, so I did it like this:
```
<Carousel ItemsSource="{Binding ItemSourcePr, ElementName=CoureselControl}" Name="carousel" Grid.Row="1">
        <Carousel.PageTransition>
          <PageSlide Duration="0.1" Orientation="Horizontal" />
        </Carousel.PageTransition>
        <Carousel.ItemTemplate>
          <DataTemplate>
            <Image Source="{Binding}" Stretch="Uniform" Width="200" Height="150"/>
          </DataTemplate>
        </Carousel.ItemTemplate>
      </Carousel>
```
But I get x:DataType error in DataTemplate tag

**Reply by vo1s** [reply]
I set to x:DataType="Bitmap"
And it works fine. But still have a lot of memory consumption problem. Maybe somehow I need to clean it? If I get 138 photos displayed I get 1gb RAM consumption  more than it was before. Maybe I need to compress it?

**Reply by thevortexcloud** [reply]
That happens if you have compiled bindings set to be used by default (which I think is the default now in the templates?). You can either use a `ReflectionBinding` or set up your view properly to use a CompiledBinding. The latter is preferable as it will give you proper compile errors for obvious problems. To do that set `x:DataType` on your view to your view model class. EG

```xaml
<Window xmlns="https://github.com/avaloniaui"
        xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
        xmlns:vm="using:My.NameSpace.ViewModels"
        xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
        xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
        mc:Ignorable="d" d:DesignWidth="800" d:DesignHeight="450"
        x:Class="My.NameSpace.Views.MyWindow"
        x:DataType="vm:MyWindowViewModel"
```
etc

**Reply by thevortexcloud** [reply]
>  I set to x:DataType="Bitmap" And it works fine. But still have a lot of memory consumption problem. Maybe somehow I need to clean it? If I get 138 photos displayed I get 1gb RAM consumption more than it was before. Maybe I need to compress it?

How are you generating the Bitmaps? I have found things like `Bitmap.DecodeToWidth` seem to use far more memory than they should.

But you should try to dispose of the images and lazy load them when they are not needed. 138 images is frankly a lot of images to be displaying in a control like this. It's really only meant for a dozen or so. More than that and it gets kind of awkward from a UX point of view. But the Carousel does use some special virtulised panel apparently by default.

**Reply by vo1s** [reply]
I do it like this:
```
var files = FileManagerService.GetPlateNumber(PathToCurrentCarFolder);
            foreach (var s in files)
            {
                MainImageCollection.Add(new Bitmap(s));
            }
```

**Reply by thevortexcloud** [reply]
You probably want to do this a different way. This is eager loading which is the least optimal way to do this. It's fine for a small number of images but once you get above ~40 it can become a problem. For this you probably want to make a dedicated view model for your images. Then declare an async method to load the images. Here is very roughly how I do it in my code (albeit using a different type of items control and with a lot of code removed):

```csharp
   public class ImageViewModel : ViewModelbase {
      /// <summary>
        /// Returns the unique filename of this instance
        /// </summary>
        public string FileName { get; init; }

        /// <summary>
        /// Returns a new instance of the thumbnail for this instance
        /// </summary>
        public Task<Bitmap?> ThumbnailAsync => this.LoadThumbnailAsync();

    public ImageViewModel(string fileName) {
            this.FileName = fileName;
        }

 /// <summary>
        /// Returns a bitmap with a small thumbnail version of the <see cref="MainImage"/>
        /// </summary>
        public Task<Bitmap?> LoadThumbnailAsync() {
return Task.Run(() => new Bitmap(this.FileName));
}
}
```
```xaml
            <ItemsRepeater.ItemTemplate>
                <DataTemplate>
                        <Image Source="{CompiledBinding ThumbnailAsync^}" Width="200" />
                </DataTemplate>
            </ItemsRepeater.ItemTemplate>
```

I also suggest resizing the images if they are fairly large. It can make a huge difference. I am lazy loading >2k images in my app and don't really get more than ~200mb of memory usage most of the time. It can get much higher though if I am looking at a very large image.

**Reply by vo1s** [reply]
I have created a small service similar to yours:
```
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Threading.Tasks;
using Avalonia.Media.Imaging;

namespace AvaloniaScantrak.Services
{
    public class LoadImageAsyncService
    {
        
        public Task<ObservableCollection<Bitmap>?> ThumbnailAsync(IEnumerable<string> fileNames)
        {
            return LoadThumbnailAsync(fileNames);
        }
        
        private async Task<ObservableCollection<Bitmap>?> LoadThumbnailAsync(IEnumerable<string> fileNames)
        {
            try
            {
                var thumbnails = new ObservableCollection<Bitmap>();

                foreach (var fileName in fileNames)
                {
                    
                        var bitmap = await LoadBitmapAsync(fileName);
                        thumbnails.Add(bitmap);
                    
                }

                return thumbnails;
            }
            catch (Exception ex)
            {
                // Handle exceptions if needed
                Console.WriteLine($"Error loading thumbnails: {ex.Message}");
                return null;
            }
        }

        private async Task<Bitmap> LoadBitmapAsync(string fileName)
        {
            return await Task.Run(() => new Bitmap(fileName));
        }
    }
}
```
And Use it like this:
```
var tops = FileManagerService.GetTopPictures(PathToCurrentCarFolder);
                Task.Run(async () =>
                {
                    MainImageCollection = await loadImageService.ThumbnailAsync(tops);
                  
                });
```
```
<Carousel ItemsSource="{CompiledBinding ItemSourcePr, ElementName=CoureselControl}" Name="carousel" Grid.Row="1">
        <Carousel.PageTransition>
          <PageSlide Duration="0.1" Orientation="Horizontal" />
        </Carousel.PageTransition>
        <Carousel.ItemTemplate>
          <DataTemplate x:DataType="Bitmap">
            <Image Source="{Binding}" Stretch="Uniform" Width="200" Height="150"/>
          </DataTemplate>
        </Carousel.ItemTemplate>
      </Carousel>
```
Is that a right way? Because memory consumption is still the same.

**Reply by timunie** [reply]
if you load thousands of images, you have all of them in memory... any images in memory are not compressed, so it is 4xhightxwidth bytes per image. You need to add some caching mechanism maybe and only load the image you want to show at the time. Check also Avalonia.Labs

**Reply by thevortexcloud** [reply, marked as answer]
> I have created a small service similar to yours:

This is functionally not very similar at all. It's still the same as what you had before. Just now it does things async. Notice how I am binding to `ThumbnailAsync^`? This does some magic which means it will only be called if the item is made real by the control. Since the control is virtualised only things that are visible are ever made real. Meaning it wont try to load every every image at once. 

My item control is simply just bound to a list of `ImageViewModel`. 

I also agree with Tim, using caching  helps a lot. But there is no point if you are not resizing images. You can worry about that however once you get this working.

**Reply by vo1s** [reply]
Thank you for your help.
